skyportal/handlers/api/image_analysis.py

```diff
@@ -0,0 +1,451 @@
+            )[0]
+            # We do not have difference image, so it will only display original one, template and mask
+            plots.plot_cutout(cutout, qq=[0.5, 99.9], stretch='linear')
+
+        # compute upper limit
+        ra0, dec0, sr0 = astrometry.get_frame_center(
+            wcs=wcs, width=image.shape[1], height=image.shape[0]
+        )
+        pixscale = astrometry.get_pixscale(wcs=wcs)
+
+        zero_fn = m[
+            'zero_fn'
+        ]  # Function to get the zero point as a function of position on the image
+        obj['mag_calib'] = obj['mag'] + zero_fn(obj['x'], obj['y'])
+
+        # We may roughly estimage the effective gain of the image from background mean and rms as gain = mean/rms**2
+        bg, rms = photometry.get_background(image, mask=mask, get_rms=True)
+
+        log(f'Effective gain is {np.median(bg/rms**2)}')
+
+        psf_model, psf_snapshots = psf.run_psfex(
+            image,
+            mask=mask,
+            checkimages=['SNAPSHOTS'],
+            order=0,
+            verbose=True,
+            _workdir=workdir_psfex,
+        )
+
+        sims = []
+        for _ in tqdm(range(100)):
+            image1 = image.copy()
+
+            # Simulate 20 random stars
+            sim = pipeline.place_random_stars(
+                image1,
+                psf_model,
+                nstars=20,
+                minflux=10,
+                maxflux=1e6,
+                wcs=wcs,
+                gain=gain,
+                saturation=50000,
+            )
+
+            sim['mag_calib'] = sim['mag'] + zero_fn(sim['x'], sim['y'])
+            sim['detected'] = False
+            sim['mag_measured'] = np.nan
+            sim['magerr_measured'] = np.nan
+            sim['flags_measured'] = np.nan
+
+            mask1 = image1 >= 50000
+
+            obj1 = photometry.get_objects_sextractor(
+                image1,
+                mask=mask | mask1,
+                r0=1,
+                aper=5.0,
+                wcs=wcs,
+                gain=gain,
+                minarea=3,
+                sn=5,
+                _workdir=workdir_sextractor_obj1,
+            )
+
+            obj1['mag_calib'] = obj1['mag'] + zero_fn(obj1['x'], obj1['y'])
+
+            # Positional match within FWHM/2 radius
+            oidx, sidx, dist = spherical_match(
+                obj1['ra'],
+                obj1['dec'],
+                sim['ra'],
+                sim['dec'],
+                pixscale * np.median(obj1['fwhm']) / 2,
+            )
+            # Mark matched stars
+            sim['detected'][sidx] = True
+            # Also store measured magnitude, its error and flags
+            sim['mag_measured'][sidx] = obj1['mag_calib'][oidx]
+            sim['magerr_measured'][sidx] = obj1['magerr'][oidx]
+            sim['flags_measured'][sidx] = obj1['flags'][oidx]
+
+            sims.append(sim)
+
+        sims = vstack(sims)
+
+        # FIXME
+        df = {
```

Comment:
Certainly not a good habit to call this non-DataFrame `df`. I think it may also be useful to explain what is happening here.